### PR TITLE
Core persistence priority rate limiting

### PR DESCRIPTION
**Pull request by yycptt**
**What changed?**
- Priority rate limiting for core persistence apis

**Why?**
- System stability

**How did you test it?**
- Tested locally.
- Adding unit test

**Potential risks**

**Is hotfix candidate?**
- Yes

### CODE REVIEW

File: common/persistence/client/quotas.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+	APIPriorityOverride = map[string]int{
+		"GetOrCreateShard": 0,
+		"UpdateShard":      0,
+
+		// this is a preprequisite for checkpoint queue process progress
+		"RangeCompleteHistoryTasks": 0,
+	}
+
+	RequestPrioritiesOrdered = []int{0, 1}
```
Comment (yiminc) [inline common/persistence/client/quotas.go:46]: Need away to make sure all priorities used above are included here.

File: common/persistence/persistenceRateLimitedClients.go [hunk 2]
```diff
@@ -147,7 +153,7 @@ func (p *shardRateLimitedPersistenceClient) GetOrCreateShard(
 	ctx context.Context,
 	request *GetOrCreateShardRequest,
 ) (*GetOrCreateShardResponse, error) {
-	if ok := p.rateLimiter.Allow(); !ok {
+	if ok := allow(ctx, "GetOrCreateShard", p.rateLimiter); !ok {
```
Comment (yiminc) [inline common/persistence/persistenceRateLimitedClients.go:156]: another candidate for auto-gen client @dnr :)

File: common/rpc/interceptor/rate_limit.go [hunk 3]
```diff
@@ -77,7 +77,8 @@ func (i *RateLimitInterceptor) Intercept(
 	if !i.rateLimiter.Allow(time.Now().UTC(), quotas.NewRequest(
 		methodName,
 		token,
-		"", // this interceptor layer does not throttle based on caller
+		"", // this interceptor layer does not throttle based on caller name
```
Comment (yiminc) [inline common/rpc/interceptor/rate_limit.go:80]: I thought you removed caller name from this PR?

File: service/matching/taskWriter.go [hunk 4]
```diff
@@ -216,6 +217,8 @@ func (w *taskWriter) appendTasks(
 }
 
 func (w *taskWriter) taskWriterLoop(ctx context.Context) error {
+	ctx = headers.SetCallerInfo(ctx, headers.NewCallerInfo(headers.CallerTypeBackground))
```
Comment (yiminc) [inline service/matching/taskWriter.go:220]: Will this cause AddTask (that is not sync matched) more likely to fail?

File: common/rpc/interceptor/rate_limit.go (hunk 3 above)
Comment (yycptt) [reply common/rpc/interceptor/rate_limit.go:80]: I removed caller from the values I set in context. 

The caller here is defined in `quotas.Request` which is existing code for frontend per namespace ratelimiting.

File: common/persistence/client/quotas.go (hunk 1 above)
Comment (yycptt) [reply common/persistence/client/quotas.go:46]: Done. Added unit test.

File: service/matching/taskWriter.go (hunk 4 above)
Comment (yycptt) [reply service/matching/taskWriter.go:220]: Double checked all matching db requests and marked all of them as background.

**Review by yiminc — APPROVED** [review]

File: common/cluster/metadata.go [hunk 5]
```diff
@@ -227,11 +227,12 @@ func (m *metadataImpl) Start() {
 		return
 	}
 
-	err := m.refreshClusterMetadata(context.Background())
+	ctx := headers.SetCallerInfo(context.TODO(), headers.NewCallerInfo(headers.CallerTypeBackground))
```
Comment (yiminc) [inline common/cluster/metadata.go:230]: use context.Background() instead of TODO() unless you want to change it later to something else.

**Review by yiminc — APPROVED** [review]

File: common/cluster/metadata.go (hunk 5 above)
Comment (yycptt) [reply common/cluster/metadata.go:230]: Yeah, I think we might want to specify some timeout for the context so used TODO() here. let me add a comment.

**Pull request merged**